src/query/v2/plan/operator.cpp

```diff
@@ -462,6 +463,202 @@ class DistributedScanAllAndFilterCursor : public Cursor {
   std::optional<std::vector<Expression *>> filter_expressions_;
 };
 
+class DistributedScanAllByPrimaryKeyCursor : public Cursor {
+ public:
+  explicit DistributedScanAllByPrimaryKeyCursor(
+      Symbol output_symbol, UniqueCursorPtr input_cursor, const char *op_name,
+      std::optional<storage::v3::LabelId> label,
+      std::optional<std::pair<storage::v3::PropertyId, Expression *>> property_expression_pair,
+      std::optional<std::vector<Expression *>> filter_expressions, std::optional<std::vector<Expression *>> primary_key)
+      : output_symbol_(output_symbol),
+        input_cursor_(std::move(input_cursor)),
+        op_name_(op_name),
+        label_(label),
+        property_expression_pair_(property_expression_pair),
+        filter_expressions_(filter_expressions),
+        primary_key_(primary_key) {
+    ResetExecutionState();
+  }
+
+  enum class State : int8_t { INITIALIZING, COMPLETED };
+
+  using VertexAccessor = accessors::VertexAccessor;
+
+  bool MakeRequest(RequestRouterInterface &request_router, ExecutionContext &context) {
+    {
+      SCOPED_REQUEST_WAIT_PROFILE;
+      std::optional<std::string> request_label = std::nullopt;
+      if (label_.has_value()) {
+        request_label = request_router.LabelToName(*label_);
+      }
+      current_batch_ = request_router.ScanVertices(request_label);
+    }
+    current_vertex_it_ = current_batch_.begin();
+    request_state_ = State::COMPLETED;
+    return !current_batch_.empty();
+  }
+
+  bool MakeRequestSingleFrame(Frame &frame, RequestRouterInterface &request_router, ExecutionContext &context) {
+    {
+      SCOPED_REQUEST_WAIT_PROFILE;
+      std::optional<std::string> request_label = std::nullopt;
+      if (label_.has_value()) {
+        request_label = request_router.LabelToName(*label_);
+      }
+
+      // Evaluate the expressions that hold the PrimaryKey.
+      ExpressionEvaluator evaluator(&frame, context.symbol_table, context.evaluation_context, context.request_router,
+                                    storage::v3::View::NEW);
+
+      std::vector<msgs::Value> pk;
+      MG_ASSERT(primary_key_);
+      for (auto *primary_key : *primary_key_) {
+        pk.push_back(TypedValueToValue(primary_key->Accept(evaluator)));
+      }
+
+      msgs::Label label = {.id = msgs::LabelId::FromUint(label_->AsUint())};
+
+      msgs::GetPropertiesRequest req = {.vertex_ids = {std::make_pair(label, pk)}};
+      auto get_prop_result = request_router.GetProperties(req);
+      MG_ASSERT(get_prop_result.size() <= 1);
+
+      if (get_prop_result.empty()) {
+        current_batch_ = std::vector<VertexAccessor>{};
+      } else {
+        auto properties = get_prop_result[0].props;
+        // TODO (gvolfing) figure out labels when relevant.
+        msgs::Vertex vertex = {.id = get_prop_result[0].vertex, .labels = {}};
+
+        current_batch_ = {VertexAccessor(vertex, properties, &request_router)};
```

Comment:
I think this is more complex then it should be. Here we don't want to deal with a batch, only a single vertex. So basically we either found the vertex or not, but we will never find more than 1 vertex. Therefore using a vector to store the result is kind of confusing.